dashboard/app/controllers/api/v1/user_school_infos_controller.rb

```diff
@@ -22,7 +22,7 @@ def update
       school_info_params[:country] = 'US'
     end
 
-    existing_school_info = current_user.last_complete_school_info
+    existing_school_info = Queries::UserSchoolInfo.by_user(current_user).last_complete&.school_info
```

Comment:
As much as I'd like us to be moving toward this pattern, I find this syntax _much_ less readable that simply having a method on User.  I'm not sure yet how to have the best of both worlds.

Reply:
Yeah, I struggled with that a bit. At least part of the problem here is that we have both `user_school_info` and `school_info` objects; the current approach of having a `last_complete_user_school_info` and a `last_complete_school_info` method looks super clean when you're looking at individual method calls, but (at least for me) makes it hard to think about what's going on when looking at the code as a whole.

One possibility would be to make the Query object more specific; the current implementation is constructed with a scope and then defines a method that uses that scope, with the idea being you could collect various related methods on this same object. But another approach would be to make the object specific to a single query; something like

```ruby
# Most recently created user_school_info referring to a complete
# school_info entry
class Queries::LastCompleteUserSchoolInfo
  def initialize(scope = UserSchoolInfo.all)
    @scope.
      includes(:school_info).
      select {|usi| usi.school_info.complete?}.
      sort_by(&:created_at).
      last
  end

  def self.by_user(user)
    new(user.user_school_infos)
  end
end
```

```ruby
existing_school_info = Queries::LastCompleteUserSchoolInfo.by_user(current_user)&.school_info
```

Reply:
But I like what you're proposing, grouping queries by relevant model (where one model is clearly more relevant than another).  It's also possible that the readability challenge of this particular case arises not only from using a query object but from various code smells in how we've structured this data in the first place. (`user_school_info -> school_info -> school` is already problematic.)  I still hold that `UserSchoolInfo` ought to be renamed to `Tenure`.

Reply:
One possible direction, but maybe this breaks a principle: Should a query object be named for the thing it returns, or for the thing we're asking questions of?  I could certainly see a world where we ask

```rb
Queries::User.new(current_user).last_complete_user_school_info
Queries::User.new(current_user).last_complete_school_info
```

...but that almost certainly drives us further from reuse.  On the other hand, we might add a class and usefully hide away the `UserSchoolInfo` layer in some cases (and reduce trainwreck-y syntax):

```rb
Queries::UserSchoolInfo.by_user(current_user).last_complete
Queries::SchoolInfo.by_user(current_user).last_complete
```

...where the one can be implemented in terms of the other, although we may not want deep dependencies between these things.

Reply:
Although this does very much turn the query object from "a place to put various bits of query logic that relate to this concept" into "a thing that does exactly this query". I don't know how I feel about which direction we'd prefer to move in

Reply:
I'm also _most_ interested in this concept for when we start getting into a multi-model query-space.  LP actually has a related discussion coming up about what our ideal query interface for progress info looks like, and I could see us building `Queries::Progress` class with `for_user` and `for_section` starter scopes, that does a much nicer job of encoding some of our model relationships than the models themselves do.

Reply:
Similarly, we could do something like

```ruby
class Queries::UserSchoolInfo
  # ...
  def last_complete_school_info
    last_complete&.school_info
  end
end

# ...

existing_school_info = Queries::UserSchoolInfo.by_user(current_user).last_complete_school_info
```

which makes for a cleaner `existing_school_info` call, but turns the `Queries::UserSchoolInfo` object into something that sometimes gives you a UserSchoolInfo and sometimes a SchoolInfo

Reply:
Here's a question: In either case
```rb
Queries::User.new(current_user).last_complete_school_info
Queries::SchoolInfo.by_user(current_user).last_complete
```
...how is a query object any better than having mixed-in module or concern providing relevant methods on the named model?
```rb
current_user.last_complete_school_info # hmm
SchoolInfo.where(user: current_user).last_complete
```
I think the answer has to do with crowding the namespace, shadowing, complexity and possible side-effects, isolation from the framework.  Although I think this is a tough one, since we're tightly bound to ActiveRecord either way.  It's possible that more verbose code is just the price we pay for separating these from the framework, where everything is by convention.  If so, that's fine.  I just want to make sure we're keeping it as usable as possible.

I wonder what thinking we need to do about discover-ability of these query objects once they're written.

Reply:
One point that Winter raised is that we should perhaps take the time to ask ourselves if reuse is actually _always_ a goal here; there might be value in having both things like the Race and Gender policies that are reusable and in having things like "last complete school info" that we don't expect we'll ever need outside the context of a single user.

Reply:
> I could see us building a Queries::Progress class with for_user and for_section starter scopes, that does a much nicer job of documenting some of our model relationships than the models themselves do.

That sounds potentially beautiful

Reply:
In the spirit of [Google's standard of code review](https://google.github.io/eng-practices/review/reviewer/standard.html) I think any of these approaches "definitely improves the overall code health" and will happily :+1: when the time comes.

Having a `Queries::UserSchoolInfo` and a `Queries::SchoolInfo` is my favorite so far, but I don't have a particularly good reason yet.

Another idea: Maybe we immediately make this a multi-model query object `Queries::Schools` and split it later if we find that's too big a bucket.

Reply:
The full-on query object pattern here gets us two things:

1. Separates querying logic into a different file (organization)
2. Ability to find the "last complete" UserSchoolInfo from arbitrary sets of UserSchoolInfos, not just ones associated with a user(flexibility / composability)

#2 is the one which is decreasing readability here in exchange for flexibility. If it's not something we want, we can still choose to get #1 by itself, in which case the function call might look like:

```
Queries::User.last_complete_school_info(current_user)
```

Reply:
Picking this back up, it does seem like the "construct the Query Object with a scope" cleverness that I was trying to do was getting in the way more than anything else. I do still see that as being a useful pattern for _something_, but not this.

I've kept an example of that pattern in the README, but updated the query object itself to just use a nice simple class method.

I do think that I prefer the idea of the query objects being named after the kind of thing they're trying to retrieve than the kind of thing they're querying against, so I've created both a `Queries::UserSchoolInfo` and a `Queries::SchoolInfo` for this case. I can definite see the value in a more-generic `Queries::School` object, but I'm hesitant to introduce a new naming convention without committing to it.

Reply:
Especially since we already have a `School` model... that could get quite confusing!